deployer/deployer-services/src/main/java/io/hyscale/deployer/services/deployer/impl/KubernetesDeployer.java

```diff
@@ -246,66 +248,53 @@ public ServiceAddress getServiceAddress(DeploymentContext context) throws Hyscal
         } catch (HyscaleException e) {
             logger.error("Error while fetching status {} ", e);
             throw e;
         }
-
-        return getPodDeploymentStatus(context, v1PodList);
+        
+        if (podParent == null) {
+            return DeploymentStatusUtil.getNotDeployedStatus(serviceName);
+        }
+        PodParentHandler podParentHandler = PodParentFactory.getHandler(podParent.getKind());
+        return updateServiceAddress(context, podParentHandler.buildStatus(podParent.getParent()));
     }
-
-    /**
-     * Get ServiceStatus for services based on app label
-     */
+    
     @Override
     public List<DeploymentStatus> getDeploymentStatus(DeploymentContext context) throws HyscaleException {
-        List<DeploymentStatus> deploymentStatusList = new ArrayList<>();
+        List<PodParent> podParentList = null;
+        String selector = ResourceSelectorUtil.getSelector(context.getAppName());
+        V1PodHandler podHandler = (V1PodHandler)ResourceHandlers.getHandlerOf(ResourceKind.POD.getKind());
         try {
             ApiClient apiClient = clientProvider.get((K8sAuthorisation) context.getAuthConfig());
-
-            Set<String> services = K8sDeployerUtil.getDeployedServices(apiClient, context);
-            for (String serviceName : services) {
-                context.setServiceName(serviceName);
-                deploymentStatusList.add(getServiceDeploymentStatus(context));
-            }
+            podParentList = podHandler.getPodParentsList(apiClient, selector, context.getNamespace());
         } catch (HyscaleException e) {
             logger.error("Error while fetching status {} ", e);
             throw e;
 
         }
-        return deploymentStatusList;
-    }
-
-    private DeploymentStatus getPodDeploymentStatus(DeploymentContext context, List<V1Pod> v1PodList) {
-
-        if (v1PodList == null || v1PodList.isEmpty()) {
+        if (podParentList == null || podParentList.isEmpty()) {
             return null;
         }
-        DeploymentStatus deploymentStatus = new DeploymentStatus();
-        deploymentStatus.setServiceName(context.getServiceName());
-        deploymentStatus.setAge(DeploymentStatusUtil.getAge(v1PodList));
-        deploymentStatus.setMessage(DeploymentStatusUtil.getMessage(v1PodList));
-        deploymentStatus.setServiceStatus(DeploymentStatusUtil.getStatus(v1PodList));
-
+        List<DeploymentStatus> deploymentStatusList =  podParentList.stream().map(each -> {
+            PodParentHandler podParentHandler = PodParentFactory.getHandler(each.getKind());
+            return podParentHandler.buildStatus(each.getParent());
+          }).collect(Collectors.toList());
+        return deploymentStatusList.stream().map(each -> updateServiceAddress(context, each)).collect(Collectors.toList());
+    }
+    
+    private DeploymentStatus updateServiceAddress(DeploymentContext context, DeploymentStatus deploymentStatus) {
```

Comment:
KuberntesDeployer itself is handling all the overhead of the service address,
can we decouple the contract and delegate it some other class like
ServiceAddressHandler ?

Reply:
ServiceAddressHandler / ServiceStatusProcessor -> for processing the deployment status .